core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/scheduling/resourcePolicies/ResourceAllocator.scala

```diff
@@ -30,27 +45,92 @@ class DefaultResourceAllocator(
-      region.getEffectiveOperators
-        .map(physicalOpId => physicalPlan.getOperator(physicalOpId))
-        .map { physicalOp =>
-          {
-            val workerCount = if (physicalOp.parallelizable) {
-              physicalOp.suggestedWorkerNum match {
-                // Keep suggested number of workers
-                case Some(num) => num
-                // If no suggested number, use default value
-                case None => AmberConfig.numWorkerPerOperatorByDefault
-              }
-            } else {
-              // Non parallelizable operator has only 1 worker
-              1
-            }
-            physicalOp.id -> (0 until workerCount).map(_ => WorkerConfig()).toList
-          }
-        }
-        .toMap
-    )
+
+    val opToWorkerConfigsMapping = region.getEffectiveOperators
+      .map(physicalOpId => physicalPlan.getOperator(physicalOpId))
+      .map(physicalOp => physicalOp.id -> generateWorkerConfigs(physicalOp))
+      .toMap
+
+    workerConfigs ++= opToWorkerConfigsMapping
+
+    // assign workers to physical plan
+    // TODO: move workers information into WorkerConfig completely
+    opToWorkerConfigsMapping.toList.foreach {
+      case (physicalOpId, workerConfigs) =>
+        physicalPlan.getOperator(physicalOpId).assignWorkers(workerConfigs.length)
+    }
+
+    propagatePartitionRequirementInRegion(region)
+
+    val linkToChannelConfigsMapping = region.getEffectiveLinks.map { physicalLinkId =>
+      physicalLinkId -> generateChannelConfigs(
+        workerConfigs.getOrElse(physicalLinkId.from, List()).map(_.workerId),
+        workerConfigs.getOrElse(physicalLinkId.to, List()).map(_.workerId),
+        outputPartitionInfos(physicalLinkId.from)
+      )
+    }.toMap
+
+    channelConfigs ++= linkToChannelConfigsMapping
+
+    val config = RegionConfig(opToWorkerConfigsMapping, linkToChannelConfigsMapping)
+
     (region.copy(config = Some(config)), 0)
   }
+
+  /**
+    * This method propagates partitioning requirements in the PhysicalPlan DAG.
+    *
+    * To be relaxed:
+    *   This method is supposed to be invoked once for each region, and only propagate partitioning requirements within
+    *   the region, however, we can only do full physical plan propagation due to link dependency from other regions.
+    *
+    *   For example, suppose we have the following physical Plan:
+    *
+    *     A ->
+    *           HJ
+    *     B ->
+    *
+    *   When propagate the first region A -> HJ, it requires the partitioning requirement of all HJ's input links,
+    *   which contains the link B-HJ from the second region.
+    *
+    * This method also applies the following optimization:
+    *  - if the upstream of the link has the same partitioning requirement as that of the downstream, and their
+    *  number of workers are equal, then the partitioning on this link can be optimized to OneToOne.
+    */
+  private def propagatePartitionRequirementInRegion(region: Region): Unit = {
```

Comment:
better break this big function into `propagatePartitionRequirementInRegion` and `enforcePartitionRequirement` as what we did before?